### PR TITLE
[Azure] Use SkyPilot provisioner for status query

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR uses SkyPilot provisioned to query the VM status for Azure, making it more robust against the original CLI-based query.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-azure-query -i 0 --cloud azure --cpus 2 echo hi`
  - [x] ``for i in `seq 1 10000`; do sky status -r test-azure-query; done``
  - [x] `sky autostop --down -i 0 test-azure-query`
  - [x] ``for i in `seq 1 10000`; do sky status -r test-azure-query; done``
  - [x] `sky launch -c test-azure-query -i 0 --cloud azure --cpus 2 --num-nodes 2 echo hi`
  - [x] ``for i in `seq 1 10000`; do sky status -r test-azure-query; done``
  - [x] `` for i in `seq 1 10`; do sky launch --cloud azure --cpus 2 --num-nodes 2 -y echo hi; done`` and `sky status -r`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/provision/azure/instance.py [hunk 1]
```diff
@@ -93,3 +97,155 @@ def cleanup_ports(
...
+                return False
+        return True
+
+    try:
+        list_virtual_machines = get_azure_sdk_function(
+            client=compute_client.virtual_machines, function_name='list')
+        vms = list_virtual_machines(resource_group_name=resource_group)
+        nodes = list(filter(match_tags, vms))
+    except azure.exceptions().ResourceNotFoundError as e:
+        if 'ResourceGroupNotFound' in str(e):
```
Comment (cblmemo) [inline sky/provision/azure/instance.py:188]: Could we have a constant for this? Seems like this is widely used in a lot of PRs including #3032 and #3697.

File: sky/provision/azure/instance.py [hunk 2]
```diff
@@ -93,3 +97,155 @@ def cleanup_ports(
...
+    }
+    provisioning_state_map = {
+        'Creating': status_lib.ClusterStatus.INIT,
+        'Updating': status_lib.ClusterStatus.INIT,
+        'Failed': status_lib.ClusterStatus.INIT,
+        'Migrating': status_lib.ClusterStatus.INIT,
+        'Deleting': None,
+        # Succeeded in provisioning state means the VM is provisioned but not
+        # necessarily running.
+        # 'Succeeded': status_lib.ClusterStatus.UP,
```
Comment (cblmemo) [inline sky/provision/azure/instance.py:222]:
```suggestion
        # Succeeded in the provisioning state means the VM is provisioned but not
        # necessarily running. We don't include this status here and further query
        # the power state of the vm to determine its status later.
        # 'Succeeded': status_lib.ClusterStatus.UP,
```

**Review by cblmemo — APPROVED** [review]
Thanks for updating this @Michaelvll ! It looks great to me 🚀 Left several nits, other than that it is ready to go!

File: sky/provision/azure/instance.py [hunk 3]
```diff
@@ -93,3 +97,155 @@ def cleanup_ports(
     # Azure will automatically cleanup network security groups when cleanup
     # resource group. So we don't need to do anything here.
     del cluster_name_on_cloud, ports, provider_config  # Unused.
+
+
+# def terminate_instances(
+#     cluster_name_on_cloud: str,
+#     provider_config: Optional[Dict[str, Any]] = None,
+#     worker_only: bool = False,
+# ) -> None:
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/provision/azure/instance.py:106]: Oops, this will be added in 3700

File: sky/provision/azure/instance.py (hunk 1 above)
Comment (Michaelvll) [reply sky/provision/azure/instance.py]: Added. Thanks!

**Pull request merged**